## 1.

Перевірте, що набір публічних параметрів $(N, p, q, d) = (7, 3, 41, 2)$ криптосистеми $NTRUCrypt$ є коректним.

$N=7$ - просте;

$d=2$ - натуральне;

$p=3$ - просте;

Перевіримо, що натуральне $q$ взаємно простое з $N$ і $p$, і таке, що $q>(6d+1)p:$ 

In [81]:
N, p, q, d = 7, 3, 41, 2

In [82]:
def gcd(a, b):
    return a if b == 0 else gcd(b, a%b)

In [83]:
if gcd(q, N) == 1:
    print('q i N взаємно прості.')
else:
    print('q i N НЕ взаємно прості.')

q i N взаємно прості.


In [84]:
if gcd(q, p) == 1:
    print('q i p взаємно прості.')
else:
    print('q i p НЕ взаємно прості.')

q i p взаємно прості.


In [85]:
print(q > (6*d + 1)*p)

True


Отже, набір публічних параметрів $(N, p, q, d) = (7, 3, 41, 2)$ криптосистеми $NTRUCrypt$ є коректним.

Використовуючи секретний ключ 

$$sk=(x^6+x^5-x^4+x^3-1,x^5+x^3+2x)$$

розшифруйте криптотекст $c=-17x^5-13x^4-20x^3-13x^2+10x+5$.

In [99]:
def mul_conv_poly(a, b, m):
    c = []
    for k in range(N):
        c_k = 0
        for i in range(N):
            for j in range(N):
                if (i + j)%N == k:
                    c_k += a[i]*b[j]
        c.append(c_k%m)
    return c

In [100]:
e = [5, 10, -13, -20, -13, -17, 0]
f = [-1, 0, 0, 1, -1, 1, 1]

In [101]:
a = mul_conv_poly(f, e, q)
a

[40, 35, 38, 36, 1, 40, 8]

$a=f*e=40+35x+38x^2+36x^3+x^4+40x^5+8x^6$

In [102]:
def center_lift(a, m):
    for i in range(N):
        while a[i] <= -m/2 or a[i] > m/2:
            if a[i] > 0:
                a[i] -= m
            else: 
                a[i] += m
    return a

In [103]:
a_center = center_lift(a, q)
a_center

[-1, -6, -3, -5, 1, -1, 8]

$a'=-1-6x-3x^2-5x^3+x^4-1x^5+8x^6$

Знайдемо $F_p$:

$x^7-1=x^7+2;\quad f=x^6+x^5-x^4+x^3-1=x^6+x^5+2x^4+x^3+2$

$x^7+2=(x+2)(x^6+x^5+2x^4+x^3+2)+2x^5+x^4+x^3+x+1$

$x^6+x^5+2x^4+x^3+2=(2x+1)(2x^5+x^4+x^3+x+1)+(2x^4+x^2+1)$

$2x^5+x^4+x^3+x+1 =(x+2)(2x^4+x^2+1)+(x^2+2)$

$2x^4+x^2+1=2x^2(x^2+2)+1$

$1=(2x^4+x^2+1)+x^2(x^2+2)=(2x^4+x^2+1)+x^2(2x^5+x^4+x^3+x+1+$

$+(2x+1)(2x^4+x^2+1))=(2x^3+x^2+1)(2x^4+x^2+1)+$

$+x^2(2x^5+x^4+x^3+x+1)=(2x^3+x^2+1)(x^6+x^5+2x^4+x^3+2+$

$+(x+2)(2x^5+x^4+x^3+x+1))+x^2(2x^5+x^4+x^3+x+1)=(2x^3+x^2+1)$

$(x^6+x^5+2x^4+x^3+2)+(2x^3+x^2+1)(x+2)(2x^5+x^4+x^3+x+1)+$

$+x^2(2x^5+x^4+x^3+x+1)=(2x^3+x^2+1)(x^6+x^5+2x^4+x^3+2)+$

$+(2x^5+x^4+x^3+x+1)(2x^4+2x^3+x+2)=(2x^3+x^2+1)(x^6+x^5+$

$+2x^4+x^3+2)+(x^7+2+(2x+1)(x^6+x^5+2x^4+x^3+2))(2x^4+$

$+2x^3+x+2)=(2x^3+x^2+1)(x^6+x^5+2x^4+x^3+2)+(x^7+2)(2x^4+$

$+2x^3+x+2)+(2x+1)(2x^4+2x^3+x+2)(x^6+x^5+2x^4+x^3+2)=$

$=(x^7+2)(2x^4+2x^3+x+2)+(x^6+x^5+2x^4+x^3+2)(x^5+x^3+2x)$

отже $F_p=x^5+x^3+2x$

In [104]:
F_p = [0, 2, 0, 1, 0, 1, 0]

In [114]:
m = center_lift(mul_conv_poly(F_p, a_center, p), p)
m

[-1, 1, 1, 1, 0, 0, -1]

Відповідь: $m=-1+x+x^2+x^3-x^6$

## 2.

Еліптична крива $E$ над полем $GF(631)$ задана рівнянням 

$$y^2=x^3+30x+34.$$ 

a) Перевірте, що точки $P=(36, 571), Q=(121, 244)$ належать кривій $E$.

In [604]:
MOD = 631
P, Q = (36, 571), (121, 244)

In [605]:
def whether_belong(point):
    x, y = point
    return (x**3 + 30*x + 34)%MOD == y**2%MOD

In [606]:
whether_belong(P)

True

In [607]:
whether_belong(Q)

True

б) Перевірте, що точки $P$ та $Q$ мають порядок $5$ і породжують підгрупу точок експоненти $5$. 

In [608]:
def inverse(a, m = MOD):
    return a**(m - 2)%m if a%m != 0 else print('\n*NOT INVERTIBLE*\n')

In [609]:
def double_point(p, m = MOD, a = 30):
    alpha = (3*p[0]**2 + a)*inverse(2*p[1], m)%m
    x = (alpha**2 - 2*p[0])%m
    y = (-p[1] - alpha*(x - p[0]))%m
    return (x, y)

In [610]:
def add_points(a, b, m = MOD):
    if a == b:
        return double_point(a, m)
    if a == (b[0], -b[1]%m) or b == (a[0], -a[1]%m):
        return (0, 0)
    if a == (0, 0):
        return b
    alpha =  (a[1] - b[1])*inverse(a[0] - b[0], m)%m
    x = (alpha**2 - a[0] - b[0])%m
    y = (-alpha*(x - a[0]) - a[1])%m
    return (x, y)

In [611]:
def point_order(point, m = MOD, n = 2):
    new_point =  double_point(point, m)
    while True:
        new_point = add_points(new_point, point, m)
        n += 1
        if new_point == (0, 0):
            return n

In [612]:
def exp_point(point, n, m = MOD, new_point = (0, 0)):
    for i in bin(n)[2:][::-1]:
        if int(i) == 1:
            new_point = add_points(new_point, point)
        point = double_point(point)
    return new_point

In [613]:
print('Порядок точки P: ', point_order(P))
print('Порядок точки Q: ', point_order(Q))

Порядок точки P:  5
Порядок точки Q:  5


Ми вже переконались, що кожна з точок $P$ і $Q$ має експоненту $5$. Покажемо тепер, що жоден степінь точки $P$ не дорівнює точці $Q$. Оскільки порядок точки $P$ дорівнює $5$, достатньо перевірити степені $2, 3, 4$, далі вони будуть зациклені:

In [614]:
for i in range(1, 11):
    print(f'{i}P = ', exp_point(P, i))

1P =  (36, 571)
2P =  (617, 5)
3P =  (617, 626)
4P =  (36, 60)
5P =  (0, 0)
6P =  (36, 571)
7P =  (617, 5)
8P =  (617, 626)
9P =  (36, 60)
10P =  (0, 0)


Таким чином, точки $P$ та $Q$ породжують підгрупу точок експоненти $5$.

в) Обчисліть значення спарювання Вейля $w_5(P,Q)$ і перевірте, що отримане значення є коренем $5$ степеня з $1$.

In [626]:
def g(p, q, m = MOD, a = 30):
    if p == (q[0], -q[1]%m) or q == (p[0], -p[1]%m):
        return lambda x, y: x + p[0]
    elif p == q:
        alpha = (3*p[0]**2 + a)*inverse(2*p[1], m)%m
    else:
        alpha =  (p[1] - q[1])*inverse(p[0] - q[0], m)%m
    return lambda x, y: (y - p[1] - alpha*(x - p[0]))*inverse(x + p[0] + q[0] - alpha**2)

In [625]:
def Miller(n, point, divisor, m = MOD):
    n = bin(n)[2:][::-1]
    t, f = point, 1
    for j in reversed(range(len(n) - 1)):
        f = f**2*g(t, t)(divisor[0], divisor[1])%m
        t = double_point(t)
        if j == 1:
            f *= g(t, point)(divisor[0], divisor[1])%m
            t = add_points(t, point)
    return f

Оберемо випадкову точку $R=(0, 36)$, що не належить підгрупі, породженій точками $P$ і $Q$. 

In [617]:
R = (0, 36)

In [618]:
def Weil_pairing(n, p, q, r, m = MOD):
    f_5P_QplusR = Miller(5, p, add_points(q, r))
    f_5P_R = Miller(5, p, r)
    f_5Q_PminusR = Miller(5, q, add_points(p, (r[0], -r[1]%m)))
    f_5Q_minusR = Miller(5, q, (r[0], -r[1]%m))
    print(f_5P_QplusR, f_5P_R, f_5Q_PminusR, f_5Q_minusR)
    return f_5P_QplusR*inverse(f_5P_R, m)*f_5Q_minusR*inverse(f_5Q_PminusR)%m

In [620]:
w = Weil_pairing(5, P, Q, R)
w

140 52 284 576


242

Користуючись алгоритмом Міллера, обчислимо значення $f_{5,P}$ і $f_{5,Q}$: 

$$f_{5,P}(Q+R)=140,$$

$$f_{5,P}(R)=52,$$

$$f_{5,Q}(P-R)=284,$$

$$f_{5,Q}(-R)=576.$$

Отже 

$$w_5(P,Q)=\frac{f_{5,P}(Q+R)}{f_{5,P}(R)}\cdot\frac{f_{5,Q}(-R)}{f_{5,Q}(P-R)}=242.$$

Перевіримо, що $w_5(P,Q)$ є корененем $5$ степеня з $1$:

In [622]:
w**5%MOD == 1

True